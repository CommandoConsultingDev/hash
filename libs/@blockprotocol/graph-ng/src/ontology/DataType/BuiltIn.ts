import * as S from "@effect/schema/Schema";

import * as Json from "../../internal/Json.js";
import { parseOrThrow } from "../DataType.js";

export const Boolean = {
  v1: parseOrThrow(
    "https://blockprotocol.org/@blockprotocol/types/data-type/boolean/v/1",
    S.boolean.pipe(S.title("Boolean"), S.description("A True or False value")),
  ),
};

export const EmptyList = {
  v1: parseOrThrow(
    "https://blockprotocol.org/@blockprotocol/types/data-type/empty-list/v/1",
    S.tuple().pipe(S.title("Empty List"), S.description("An Empty List")),
  ),
};

export const Null = {
  v1: parseOrThrow(
    "https://blockprotocol.org/@blockprotocol/types/data-type/null/v/1",
    S.null.pipe(
      S.title("Null"),
      S.description("A placeholder value representing 'nothing'"),
    ),
  ),
};

export const Number = {
  v1: parseOrThrow(
    "https://blockprotocol.org/@blockprotocol/types/data-type/number/v/1",
    S.number.pipe(
      S.title("Number"),
      S.description("An arithmetical value (in the Real number system)"),
    ),
  ),
};

export const Opaque = {
  v1: parseOrThrow(
    "https://blockprotocol.org/@blockprotocol/types/data-type/object/v/1",
    S.record(S.string, Json.Value).pipe(
      S.title("Object"),
      S.description("An opaque, untyped JSON object"),
    ),
  ),
};

export const Text = {
  v1: parseOrThrow(
    "https://blockprotocol.org/@blockprotocol/types/data-type/text/v/1",
    S.string.pipe(
      S.title("Text"),
      S.description("An ordered sequence of characters"),
    ),
  ),
};
