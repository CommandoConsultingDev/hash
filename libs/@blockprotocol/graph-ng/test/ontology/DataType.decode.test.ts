import { Either } from "effect";
import { describe, expect, test } from "vitest";

import * as DataType from "../../src/ontology/DataType.js";
import { DataTypeUrl } from "../../src/ontology/index.js";

describe("literal", () => {
  test("`number`", () => {
    const schema = {
      $id: DataTypeUrl.parseOrThrow("https://example.com/number-123/v/1"),
      $schema:
        "https://blockprotocol.org/types/modules/graph/0.3/schema/data-type",
      const: 123,
      kind: "dataType",
      title: "A constant number of value `123`",
      type: "number",
    } satisfies DataType.Schema;

    const result = DataType.fromSchema(schema);
    const dataType = Either.getOrThrow(result);

    expect(dataType).toMatchSnapshot();
  });

  test("`string`", () => {
    const schema = {
      $id: DataTypeUrl.parseOrThrow("https://example.com/string-abc/v/1"),
      $schema:
        "https://blockprotocol.org/types/modules/graph/0.3/schema/data-type",
      const: "abc",
      kind: "dataType",
      title: "A constant string of value `abc`",
      type: "string",
    } satisfies DataType.Schema;

    const result = DataType.fromSchema(schema);
    const dataType = Either.getOrThrow(result);

    expect(dataType).toMatchSnapshot();
  });

  test("`boolean`", () => {
    const schema = {
      $id: DataTypeUrl.parseOrThrow("https://example.com/boolean-true/v/1"),
      $schema:
        "https://blockprotocol.org/types/modules/graph/0.3/schema/data-type",
      const: true,
      kind: "dataType",
      title: "A constant boolean of value `true`",
      type: "boolean",
    } satisfies DataType.Schema;

    const result = DataType.fromSchema(schema);
    const dataType = Either.getOrThrow(result);

    expect(dataType).toMatchSnapshot();
  });

  test("`null`", () => {
    const schema = {
      $id: DataTypeUrl.parseOrThrow("https://example.com/null/v/1"),
      $schema:
        "https://blockprotocol.org/types/modules/graph/0.3/schema/data-type",
      const: null,
      kind: "dataType",
      title: "A constant null value",
      type: "null",
    } satisfies DataType.Schema;

    const result = DataType.fromSchema(schema);
    const dataType = Either.getOrThrow(result);

    expect(dataType).toMatchSnapshot();
  });
});

describe("string", () => {
  const base = {
    $id: DataTypeUrl.parseOrThrow("https://example.com/string/v/1"),
    $schema:
      "https://blockprotocol.org/types/modules/graph/0.3/schema/data-type",
    description: "a string",
    kind: "dataType",
    title: "string",
    type: "string",
  } satisfies DataType.Schema;

  test("standard", () => {
    const schema = base;
    const result = DataType.fromSchema(schema);
    const dataType = Either.getOrThrow(result);

    expect(dataType).toMatchSnapshot();
  });

  test("minLength", () => {
    const schema = {
      ...base,
      minLength: 1,
    } satisfies DataType.Schema;
    const result = DataType.fromSchema(schema);
    const dataType = Either.getOrThrow(result);

    expect(dataType).toMatchSnapshot();
  });

  test("maxLength", () => {
    const schema = {
      ...base,
      maxLength: 10,
    } satisfies DataType.Schema;
    const result = DataType.fromSchema(schema);
    const dataType = Either.getOrThrow(result);

    expect(dataType).toMatchSnapshot();
  });

  test("pattern", () => {
    const schema = {
      ...base,
      pattern: "^[a-z]+$",
    } satisfies DataType.Schema;
    const result = DataType.fromSchema(schema);
    const dataType = Either.getOrThrow(result);

    expect(dataType).toMatchSnapshot();
  });

  test("minLength, maxLength, pattern", () => {
    const schema = {
      ...base,
      minLength: 1,
      maxLength: 10,
      pattern: "^[a-z]+$",
    } satisfies DataType.Schema;
    const result = DataType.fromSchema(schema);
    const dataType = Either.getOrThrow(result);

    expect(dataType).toMatchSnapshot();
  });
});

describe("number", () => {
  const base = {
    $id: DataTypeUrl.parseOrThrow("https://example.com/number/v/1"),
    $schema:
      "https://blockprotocol.org/types/modules/graph/0.3/schema/data-type",
    description: "a number",
    kind: "dataType",
    title: "number",
    type: "number",
  } satisfies DataType.Schema;

  test("number", () => {
    const schema = base;
    const result = DataType.fromSchema(schema);
    const dataType = Either.getOrThrow(result);

    expect(dataType).toMatchSnapshot();
  });

  test("integer", () => {
    const schema = {
      ...base,
      type: "integer",
    } satisfies DataType.Schema;
    const result = DataType.fromSchema(schema);
    const dataType = Either.getOrThrow(result);

    expect(dataType).toMatchSnapshot();
  });

  test("multipleOf", () => {
    const schema = {
      ...base,
      multipleOf: 2,
    } satisfies DataType.Schema;
    const result = DataType.fromSchema(schema);
    const dataType = Either.getOrThrow(result);

    expect(dataType).toMatchSnapshot();
  });

  test("minimum", () => {
    const schema = {
      ...base,
      minimum: 0,
    } satisfies DataType.Schema;
    const result = DataType.fromSchema(schema);
    const dataType = Either.getOrThrow(result);

    expect(dataType).toMatchSnapshot();
  });

  test("maximum", () => {
    const schema = {
      ...base,
      maximum: 100,
    } satisfies DataType.Schema;
    const result = DataType.fromSchema(schema);
    const dataType = Either.getOrThrow(result);

    expect(dataType).toMatchSnapshot();
  });

  test("exclusiveMinimum", () => {
    const schema = {
      ...base,
      exclusiveMinimum: 0,
    } satisfies DataType.Schema;
    const result = DataType.fromSchema(schema);
    const dataType = Either.getOrThrow(result);

    expect(dataType).toMatchSnapshot();
  });

  test("exclusiveMaximum", () => {
    const schema = {
      ...base,
      exclusiveMaximum: 100,
    } satisfies DataType.Schema;
    const result = DataType.fromSchema(schema);
    const dataType = Either.getOrThrow(result);

    expect(dataType).toMatchSnapshot();
  });

  test("integer, multipleOf, minimum, maximum, exclusiveMinimum, exclusiveMaximum", () => {
    const schema = {
      ...base,
      type: "integer",
      multipleOf: 2,
      minimum: 0,
      maximum: 100,
      exclusiveMinimum: 0,
      exclusiveMaximum: 100,
    } satisfies DataType.Schema;
    const result = DataType.fromSchema(schema);
    const dataType = Either.getOrThrow(result);

    expect(dataType).toMatchSnapshot();
  });
});

test("boolean", () => {
  const schema = {
    $id: DataTypeUrl.parseOrThrow("https://example.com/boolean/v/1"),
    $schema:
      "https://blockprotocol.org/types/modules/graph/0.3/schema/data-type",
    description: "a boolean",
    kind: "dataType",
    title: "boolean",
    type: "boolean",
  } satisfies DataType.Schema;

  const result = DataType.fromSchema(schema);
  const dataType = Either.getOrThrow(result);

  expect(dataType).toMatchSnapshot();
});

test("null", () => {
  const schema = {
    $id: DataTypeUrl.parseOrThrow("https://example.com/null/v/1"),
    $schema:
      "https://blockprotocol.org/types/modules/graph/0.3/schema/data-type",
    description: "a null",
    kind: "dataType",
    title: "null",
    type: "null",
  } satisfies DataType.Schema;

  const result = DataType.fromSchema(schema);
  const dataType = Either.getOrThrow(result);

  expect(dataType).toMatchSnapshot();
});

test("array", () => {
  const schema = {
    $id: DataTypeUrl.parseOrThrow("https://example.com/array/v/1"),
    $schema:
      "https://blockprotocol.org/types/modules/graph/0.3/schema/data-type",
    description: "an array",
    kind: "dataType",
    title: "array",
    type: "array",
    const: [],
  } satisfies DataType.Schema;

  const result = DataType.fromSchema(schema);
  const dataType = Either.getOrThrow(result);

  expect(dataType).toMatchSnapshot();
});

test("object", () => {
  const schema = {
    $id: DataTypeUrl.parseOrThrow("https://example.com/object/v/1"),
    $schema:
      "https://blockprotocol.org/types/modules/graph/0.3/schema/data-type",
    description: "an object",
    kind: "dataType",
    title: "object",
    type: "object",
  } satisfies DataType.Schema;

  const result = DataType.fromSchema(schema);
  const dataType = Either.getOrThrow(result);

  expect(dataType).toMatchSnapshot();
});
