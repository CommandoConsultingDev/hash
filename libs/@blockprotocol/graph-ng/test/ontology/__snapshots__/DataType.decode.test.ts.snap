// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`array 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/array/v/1",
  "schema": {
    "_tag": "TupleType",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Description)": "an array",
      "Symbol(@effect/schema/annotation/Title)": "array",
    },
    "elements": [],
    "isReadonly": true,
    "rest": [],
  },
}
`;

exports[`boolean 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/boolean/v/1",
  "schema": {
    "_tag": "BooleanKeyword",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Description)": "a boolean",
      "Symbol(@effect/schema/annotation/Title)": "boolean",
    },
  },
}
`;

exports[`literal > \`boolean\` 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/boolean-true/v/1",
  "schema": {
    "_tag": "Literal",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Title)": "A constant boolean of value \`true\`",
    },
    "literal": true,
  },
}
`;

exports[`literal > \`null\` 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/null/v/1",
  "schema": {
    "_tag": "Literal",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Title)": "A constant null value",
    },
    "literal": null,
  },
}
`;

exports[`literal > \`number\` 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/number-123/v/1",
  "schema": {
    "_tag": "Literal",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Title)": "A constant number of value \`123\`",
    },
    "literal": 123,
  },
}
`;

exports[`literal > \`string\` 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/string-abc/v/1",
  "schema": {
    "_tag": "Literal",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Title)": "A constant string of value \`abc\`",
    },
    "literal": "abc",
  },
}
`;

exports[`null 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/null/v/1",
  "schema": {
    "_tag": "Literal",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Description)": "a null",
      "Symbol(@effect/schema/annotation/Identifier)": "null",
      "Symbol(@effect/schema/annotation/Title)": "null",
    },
    "literal": null,
  },
}
`;

exports[`number > exclusiveMaximum 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/number/v/1",
  "schema": {
    "_tag": "Refinement",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Description)": "a number",
      "Symbol(@effect/schema/annotation/JSONSchema)": {
        "exclusiveMaximum": 100,
      },
      "Symbol(@effect/schema/annotation/Title)": "number",
      "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/LessThan),
    },
    "from": {
      "_tag": "NumberKeyword",
      "annotations": {
        "Symbol(@effect/schema/annotation/Description)": "a number",
        "Symbol(@effect/schema/annotation/Title)": "number",
      },
    },
  },
}
`;

exports[`number > exclusiveMinimum 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/number/v/1",
  "schema": {
    "_tag": "Refinement",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Description)": "a number",
      "Symbol(@effect/schema/annotation/JSONSchema)": {
        "exclusiveMinimum": 0,
      },
      "Symbol(@effect/schema/annotation/Title)": "number",
      "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/GreaterThan),
    },
    "from": {
      "_tag": "NumberKeyword",
      "annotations": {
        "Symbol(@effect/schema/annotation/Description)": "a number",
        "Symbol(@effect/schema/annotation/Title)": "number",
      },
    },
  },
}
`;

exports[`number > integer 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/number/v/1",
  "schema": {
    "_tag": "Refinement",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Description)": "a number",
      "Symbol(@effect/schema/annotation/Identifier)": "Int",
      "Symbol(@effect/schema/annotation/JSONSchema)": {
        "type": "integer",
      },
      "Symbol(@effect/schema/annotation/Title)": "number",
      "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/Int),
    },
    "from": {
      "_tag": "NumberKeyword",
      "annotations": {
        "Symbol(@effect/schema/annotation/Description)": "a number",
        "Symbol(@effect/schema/annotation/Title)": "number",
      },
    },
  },
}
`;

exports[`number > integer, multipleOf, minimum, maximum, exclusiveMinimum, exclusiveMaximum 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/number/v/1",
  "schema": {
    "_tag": "Refinement",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Description)": "a number",
      "Symbol(@effect/schema/annotation/Identifier)": "Int",
      "Symbol(@effect/schema/annotation/JSONSchema)": {
        "type": "integer",
      },
      "Symbol(@effect/schema/annotation/Title)": "number",
      "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/Int),
    },
    "from": {
      "_tag": "Refinement",
      "annotations": {
        "Symbol(@effect/schema/annotation/Description)": "a number less than 100",
        "Symbol(@effect/schema/annotation/JSONSchema)": {
          "exclusiveMaximum": 100,
        },
        "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/LessThan),
      },
      "from": {
        "_tag": "Refinement",
        "annotations": {
          "Symbol(@effect/schema/annotation/Description)": "a positive number",
          "Symbol(@effect/schema/annotation/JSONSchema)": {
            "exclusiveMinimum": 0,
          },
          "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/GreaterThan),
        },
        "from": {
          "_tag": "Refinement",
          "annotations": {
            "Symbol(@effect/schema/annotation/Description)": "a number less than or equal to 100",
            "Symbol(@effect/schema/annotation/JSONSchema)": {
              "maximum": 100,
            },
            "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/LessThanOrEqualTo),
          },
          "from": {
            "_tag": "Refinement",
            "annotations": {
              "Symbol(@effect/schema/annotation/Description)": "a non-negative number",
              "Symbol(@effect/schema/annotation/JSONSchema)": {
                "minimum": 0,
              },
              "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/GreaterThanOrEqualTo),
            },
            "from": {
              "_tag": "Refinement",
              "annotations": {
                "Symbol(@effect/schema/annotation/Description)": "a number divisible by 2",
                "Symbol(@effect/schema/annotation/JSONSchema)": {
                  "multipleOf": 2,
                },
                "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/MultipleOf),
              },
              "from": {
                "_tag": "NumberKeyword",
                "annotations": {
                  "Symbol(@effect/schema/annotation/Description)": "a number",
                  "Symbol(@effect/schema/annotation/Title)": "number",
                },
              },
            },
          },
        },
      },
    },
  },
}
`;

exports[`number > maximum 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/number/v/1",
  "schema": {
    "_tag": "Refinement",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Description)": "a number",
      "Symbol(@effect/schema/annotation/JSONSchema)": {
        "maximum": 100,
      },
      "Symbol(@effect/schema/annotation/Title)": "number",
      "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/LessThanOrEqualTo),
    },
    "from": {
      "_tag": "NumberKeyword",
      "annotations": {
        "Symbol(@effect/schema/annotation/Description)": "a number",
        "Symbol(@effect/schema/annotation/Title)": "number",
      },
    },
  },
}
`;

exports[`number > minimum 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/number/v/1",
  "schema": {
    "_tag": "Refinement",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Description)": "a number",
      "Symbol(@effect/schema/annotation/JSONSchema)": {
        "minimum": 0,
      },
      "Symbol(@effect/schema/annotation/Title)": "number",
      "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/GreaterThanOrEqualTo),
    },
    "from": {
      "_tag": "NumberKeyword",
      "annotations": {
        "Symbol(@effect/schema/annotation/Description)": "a number",
        "Symbol(@effect/schema/annotation/Title)": "number",
      },
    },
  },
}
`;

exports[`number > multipleOf 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/number/v/1",
  "schema": {
    "_tag": "Refinement",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Description)": "a number",
      "Symbol(@effect/schema/annotation/JSONSchema)": {
        "multipleOf": 2,
      },
      "Symbol(@effect/schema/annotation/Title)": "number",
      "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/MultipleOf),
    },
    "from": {
      "_tag": "NumberKeyword",
      "annotations": {
        "Symbol(@effect/schema/annotation/Description)": "a number",
        "Symbol(@effect/schema/annotation/Title)": "number",
      },
    },
  },
}
`;

exports[`number > number 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/number/v/1",
  "schema": {
    "_tag": "NumberKeyword",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Description)": "a number",
      "Symbol(@effect/schema/annotation/Title)": "number",
    },
  },
}
`;

exports[`object 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/object/v/1",
  "schema": {
    "_tag": "TypeLiteral",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Description)": "an object",
      "Symbol(@effect/schema/annotation/Title)": "object",
    },
    "indexSignatures": [
      {
        "isReadonly": true,
        "parameter": {
          "_tag": "StringKeyword",
          "annotations": {
            "Symbol(@effect/schema/annotation/Description)": "a string",
            "Symbol(@effect/schema/annotation/Title)": "string",
          },
        },
        "type": {
          "_tag": "Union",
          "annotations": {
            "Symbol(@effect/schema/annotation/Type)": Symbol(@blockprotocol/graph/Json/Value),
          },
          "types": [
            {
              "_tag": "TypeLiteral",
              "annotations": {},
              "indexSignatures": [
                {
                  "isReadonly": true,
                  "parameter": {
                    "_tag": "StringKeyword",
                    "annotations": {
                      "Symbol(@effect/schema/annotation/Description)": "a string",
                      "Symbol(@effect/schema/annotation/Title)": "string",
                    },
                  },
                  "type": {
                    "_tag": "Suspend",
                    "annotations": {},
                    "ast": {
                      "_tag": "Union",
                      "annotations": {
                        "Symbol(@effect/schema/annotation/Type)": Symbol(@blockprotocol/graph/Json/Value),
                      },
                      "types": [
                        {
                          "_tag": "TypeLiteral",
                          "annotations": {},
                          "indexSignatures": [
                            {
                              "isReadonly": true,
                              "parameter": {
                                "_tag": "StringKeyword",
                                "annotations": {
                                  "Symbol(@effect/schema/annotation/Description)": "a string",
                                  "Symbol(@effect/schema/annotation/Title)": "string",
                                },
                              },
                              "type": {
                                "_tag": "Suspend",
                              },
                            },
                          ],
                          "propertySignatures": [],
                        },
                        {
                          "_tag": "TupleType",
                          "annotations": {},
                          "elements": [],
                          "isReadonly": true,
                          "rest": [
                            {
                              "_tag": "Suspend",
                            },
                          ],
                        },
                        {
                          "_tag": "StringKeyword",
                          "annotations": {
                            "Symbol(@effect/schema/annotation/Description)": "a string",
                            "Symbol(@effect/schema/annotation/Title)": "string",
                          },
                        },
                        {
                          "_tag": "NumberKeyword",
                          "annotations": {
                            "Symbol(@effect/schema/annotation/Description)": "a number",
                            "Symbol(@effect/schema/annotation/Title)": "number",
                          },
                        },
                        {
                          "_tag": "BooleanKeyword",
                          "annotations": {
                            "Symbol(@effect/schema/annotation/Description)": "a boolean",
                            "Symbol(@effect/schema/annotation/Title)": "boolean",
                          },
                        },
                        {
                          "_tag": "Literal",
                          "annotations": {
                            "Symbol(@effect/schema/annotation/Identifier)": "null",
                          },
                          "literal": null,
                        },
                      ],
                    },
                  },
                },
              ],
              "propertySignatures": [],
            },
            {
              "_tag": "TupleType",
              "annotations": {},
              "elements": [],
              "isReadonly": true,
              "rest": [
                {
                  "_tag": "Suspend",
                  "annotations": {},
                  "ast": {
                    "_tag": "Union",
                    "annotations": {
                      "Symbol(@effect/schema/annotation/Type)": Symbol(@blockprotocol/graph/Json/Value),
                    },
                    "types": [
                      {
                        "_tag": "TypeLiteral",
                        "annotations": {},
                        "indexSignatures": [
                          {
                            "isReadonly": true,
                            "parameter": {
                              "_tag": "StringKeyword",
                              "annotations": {
                                "Symbol(@effect/schema/annotation/Description)": "a string",
                                "Symbol(@effect/schema/annotation/Title)": "string",
                              },
                            },
                            "type": {
                              "_tag": "Suspend",
                            },
                          },
                        ],
                        "propertySignatures": [],
                      },
                      {
                        "_tag": "TupleType",
                        "annotations": {},
                        "elements": [],
                        "isReadonly": true,
                        "rest": [
                          {
                            "_tag": "Suspend",
                          },
                        ],
                      },
                      {
                        "_tag": "StringKeyword",
                        "annotations": {
                          "Symbol(@effect/schema/annotation/Description)": "a string",
                          "Symbol(@effect/schema/annotation/Title)": "string",
                        },
                      },
                      {
                        "_tag": "NumberKeyword",
                        "annotations": {
                          "Symbol(@effect/schema/annotation/Description)": "a number",
                          "Symbol(@effect/schema/annotation/Title)": "number",
                        },
                      },
                      {
                        "_tag": "BooleanKeyword",
                        "annotations": {
                          "Symbol(@effect/schema/annotation/Description)": "a boolean",
                          "Symbol(@effect/schema/annotation/Title)": "boolean",
                        },
                      },
                      {
                        "_tag": "Literal",
                        "annotations": {
                          "Symbol(@effect/schema/annotation/Identifier)": "null",
                        },
                        "literal": null,
                      },
                    ],
                  },
                },
              ],
            },
            {
              "_tag": "StringKeyword",
              "annotations": {
                "Symbol(@effect/schema/annotation/Description)": "a string",
                "Symbol(@effect/schema/annotation/Title)": "string",
              },
            },
            {
              "_tag": "NumberKeyword",
              "annotations": {
                "Symbol(@effect/schema/annotation/Description)": "a number",
                "Symbol(@effect/schema/annotation/Title)": "number",
              },
            },
            {
              "_tag": "BooleanKeyword",
              "annotations": {
                "Symbol(@effect/schema/annotation/Description)": "a boolean",
                "Symbol(@effect/schema/annotation/Title)": "boolean",
              },
            },
            {
              "_tag": "Literal",
              "annotations": {
                "Symbol(@effect/schema/annotation/Identifier)": "null",
              },
              "literal": null,
            },
          ],
        },
      },
    ],
    "propertySignatures": [],
  },
}
`;

exports[`string > maxLength 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/string/v/1",
  "schema": {
    "_tag": "Refinement",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Description)": "a string",
      "Symbol(@effect/schema/annotation/JSONSchema)": {
        "maxLength": 10,
      },
      "Symbol(@effect/schema/annotation/Title)": "string",
      "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/MaxLength),
    },
    "from": {
      "_tag": "StringKeyword",
      "annotations": {
        "Symbol(@effect/schema/annotation/Description)": "a string",
        "Symbol(@effect/schema/annotation/Title)": "string",
      },
    },
  },
}
`;

exports[`string > minLength 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/string/v/1",
  "schema": {
    "_tag": "Refinement",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Description)": "a string",
      "Symbol(@effect/schema/annotation/JSONSchema)": {
        "minLength": 1,
      },
      "Symbol(@effect/schema/annotation/Title)": "string",
      "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/MinLength),
    },
    "from": {
      "_tag": "StringKeyword",
      "annotations": {
        "Symbol(@effect/schema/annotation/Description)": "a string",
        "Symbol(@effect/schema/annotation/Title)": "string",
      },
    },
  },
}
`;

exports[`string > minLength, maxLength, pattern 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/string/v/1",
  "schema": {
    "_tag": "Refinement",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/ArbitraryHookId)": [Function],
      "Symbol(@effect/schema/TypeId/Pattern)": {
        "regex": /\\^\\[a-z\\]\\+\\$/,
      },
      "Symbol(@effect/schema/annotation/Description)": "a string",
      "Symbol(@effect/schema/annotation/JSONSchema)": {
        "pattern": "^[a-z]+$",
      },
      "Symbol(@effect/schema/annotation/Title)": "string",
      "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/Pattern),
    },
    "from": {
      "_tag": "Refinement",
      "annotations": {
        "Symbol(@effect/schema/annotation/Description)": "a string at most 10 character(s) long",
        "Symbol(@effect/schema/annotation/JSONSchema)": {
          "maxLength": 10,
        },
        "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/MaxLength),
      },
      "from": {
        "_tag": "Refinement",
        "annotations": {
          "Symbol(@effect/schema/annotation/Description)": "a string at least 1 character(s) long",
          "Symbol(@effect/schema/annotation/JSONSchema)": {
            "minLength": 1,
          },
          "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/MinLength),
        },
        "from": {
          "_tag": "StringKeyword",
          "annotations": {
            "Symbol(@effect/schema/annotation/Description)": "a string",
            "Symbol(@effect/schema/annotation/Title)": "string",
          },
        },
      },
    },
  },
}
`;

exports[`string > pattern 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/string/v/1",
  "schema": {
    "_tag": "Refinement",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/ArbitraryHookId)": [Function],
      "Symbol(@effect/schema/TypeId/Pattern)": {
        "regex": /\\^\\[a-z\\]\\+\\$/,
      },
      "Symbol(@effect/schema/annotation/Description)": "a string",
      "Symbol(@effect/schema/annotation/JSONSchema)": {
        "pattern": "^[a-z]+$",
      },
      "Symbol(@effect/schema/annotation/Title)": "string",
      "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/Pattern),
    },
    "from": {
      "_tag": "StringKeyword",
      "annotations": {
        "Symbol(@effect/schema/annotation/Description)": "a string",
        "Symbol(@effect/schema/annotation/Title)": "string",
      },
    },
  },
}
`;

exports[`string > standard 1`] = `
{
  "_id": "DataType",
  "id": "https://example.com/string/v/1",
  "schema": {
    "_tag": "StringKeyword",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Description)": "a string",
      "Symbol(@effect/schema/annotation/Title)": "string",
    },
  },
}
`;
